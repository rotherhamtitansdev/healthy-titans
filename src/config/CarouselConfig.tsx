
export const HomePageCarouselResponsiveConfig = {
    desktop: {
        breakpoint: { max: 3000, min: 1400 },
        items: 4,
        partialVisibilityGutter: 40,
    },
    smallDesktop: {
        breakpoint: { max: 1400, min: 1024 },
        items: 3,
        partialVisibilityGutter: 50,
    },
    tablet: {
        breakpoint: { max: 1024, min: 800 },
        items: 2,
        partialVisibilityGutter: 100,
    },
    largeMobile: {
        breakpoint: { max: 800, min: 600 },
        items: 2,
        partialVisibilityGutter: 30,
    },
    mobile: {
        breakpoint: { max: 600, min: 464 },
        items: 1,
        partialVisibilityGutter: 200,
    },
    smallMobile: {
        breakpoint: { max: 464, min: 0 },
        items: 1,
        partialVisibilityGutter: 100,
    },
};

export const FoodDetailsCarouselResponsiveConfig = {
    desktop: {
        breakpoint: { max: 3000, min: 1400 },
        items: 3,
    },
    smallDesktop: {
        breakpoint: { max: 1400, min: 1024 },
        items: 3,
    },
    tablet: {
        breakpoint: { max: 1024, min: 800 },
        items: 2,
        partialVisibilityGutter: 100,
    },
    largeMobile: {
        breakpoint: { max: 800, min: 600 },
        items: 2,
        partialVisibilityGutter: 30,
    },
    mobile: {
        breakpoint: { max: 600, min: 464 },
        items: 1,
        partialVisibilityGutter: 200,
    },
    smallMobile: {
        breakpoint: { max: 464, min: 0 },
        items: 1,
        partialVisibilityGutter: 100,
    },
};
