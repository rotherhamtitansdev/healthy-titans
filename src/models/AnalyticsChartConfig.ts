import { NHSNutritionalDataModel } from "./NHSNutritionalDataModel";

// This is the numerical instance of the NHS Nutritional guidelines model
export const NHSNutritionalBoundaries: NHSNutritionalDataModel[] = [
  { name: "Fat", low: 3, high: 17.5 },
  { name: "Saturates", low: 1.5, high: 5 },
  { name: "Sugars", low: 5, high: 22.5 },
  { name: "Salt", low: 0.1, high: 0.6 },
];

// These are the config parameters for the chart styling and sizing
export const chartSizing = {
  chartHeight: 500,
  barWidth: 30,
  CStartPos: 50,
  RStartPos: 60,
  circleYDif: 100,
  smallerRounded: 15,
};

export enum ChartLabel {
  Calories = "ENERC_KCAL",
  Protein = "PROCNT",
  Fat = "FAT",
  Saturates = "FASAT",
  Sugars = "SUGAR",
  Salt = "NA",
}

export enum Colour {
  Green = 1,
  Yellow = 2,
  Red = 3,
}
