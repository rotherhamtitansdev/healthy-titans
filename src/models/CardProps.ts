export interface CardProps {
  name: string;
  additionalStyling?: string;
  onClick?: () => void;
}
