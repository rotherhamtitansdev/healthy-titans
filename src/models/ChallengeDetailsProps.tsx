export interface ChallengeDetailsProps {
  name: string;
  description: string;
  img: string;
  rules: string[];
  equipment: string[];
  howToPlay?: string[];
}
