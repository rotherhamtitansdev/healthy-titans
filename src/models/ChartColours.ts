export interface ChartColours {
  name: string;
  primary: string;
  secondary: string;
  status: number;
  position: number;
}
