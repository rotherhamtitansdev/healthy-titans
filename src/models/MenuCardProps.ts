export interface MenuCardProps {
  key: number;
  name: string;
  description?: string;
  img?: string;
  path: string;
  externalPath?: boolean;
  disableOnClick?: boolean;
  firebaseName?: string;
  additionalStyling?: string;
}
