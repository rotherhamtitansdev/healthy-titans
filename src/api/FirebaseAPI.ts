import { getDownloadURL, getStorage, ref } from "firebase/storage";
import {
  collection,
  doc,
  getDoc,
  getDocs,
  query,
  where,
  documentId,
} from "firebase/firestore";
import { FoodDetailsProps } from "../data/nutritional_information/FoodDetailsComponentData";
import { fStore } from "../config/firebase-config";
import { NutritionalDetailsFirebaseProps } from "../models/NutritionDetailsComponentData";

interface SeeNextProps {
  key: number;
  name: string;
  path: string;
  img: string;
}

class FirebaseAPI {
  static fetchImages = async (firebaseName: string): Promise<string> => {
    const storage = getStorage();
    const starsRef = ref(storage, firebaseName);
    return getDownloadURL(starsRef);
  };

  static fetchAllImages = async (firebaseNames: string[]) =>
    Promise.all(firebaseNames.map((firebaseName) => FirebaseAPI.fetchImages(firebaseName)));

  static fetchFoodDetailsComponentsData = async (): Promise<
    FoodDetailsProps[] | undefined
  > => {
    const querySnapshot = await getDocs(collection(fStore, "FYPData"));
    if (!querySnapshot) return undefined;
     return querySnapshot.docs.map((each) => each.data() as FoodDetailsProps);
  };

  static fetchFoodDetailsSingle = async (name: string) => {
    const docRef = doc(fStore, "FYPData", name);
    const docSnap = await getDoc(docRef);
    if (docSnap) return docSnap.data() as FoodDetailsProps;
    return undefined;
  };

  static fetchFoodDetailsSeeNext = async (
    category: string,
    currentName: string
  ): Promise<{ cardData: FoodDetailsProps[]; paths: string[] }> => {
    const q = query(
      collection(fStore, "FYPData"),
      where("category", "==", category),
      where("name", "!=", currentName)
    );
    const docSnap = await getDocs(q);
    const docs = docSnap.docs.map((data) => {
      const newDoc = data.data();
      newDoc.path = data.id;
      return newDoc;
    });

    const shuffled = docs.sort(() => 0.5 - Math.random()).slice(0, 3);

    return {
      cardData: shuffled.map((value) => ({
        name: value.name,
        description: value.description,
        firebaseName: value.firebaseName,
        category: value.category,
        score: value.score,
        facts: value.facts,
      })),
      paths: shuffled.map((value) => value.path),
    };
  };

  static fetchNutritionData = async (name: string): Promise<NutritionalDetailsFirebaseProps> => {
    const querySnapshot = await getDocs(collection(fStore, "NutritionData", name, "Content"));

    const sorted = querySnapshot.docs.sort((a, b) => {
      if (a.data().order > b.data().order) {
        return 1;
      }
      if (b.data().order > a.data().order) {
        return -1;
      }
      return 0;
    });

    return sorted.map((doc2) => {
      const data = doc2.data();
      return Object.entries(data).map(([key,value]) => ({key,value}));
    });
  };

  static fetchNutritionSeeNext = async (currentName: string): Promise<SeeNextProps[]> => {
    const q = query(collection(fStore, "NutritionData"), where(documentId(), "!=", currentName));
    const querySnapshot = await getDocs(q);

    const docs = querySnapshot.docs.sort(() => 0.5 - Math.random()).slice(0, 3);

    return Promise.all(
      docs.map(async (sortedDoc, index) => {
        const dataDoc = await getDoc(doc(fStore, "NutritionData", sortedDoc.id, "Content", "Main"));

        const URI = await FirebaseAPI.fetchImages(dataDoc.data()?.firebaseName);

        return { key: index, name: dataDoc.data()?.name, path: docs[index].id, img: URI };
      })
    );
  };

  static fetchQuizData = async () => {
    const docRef = doc(fStore, "QuizData", "Quiz");
    const quizDoc = await getDoc(docRef);

    if (quizDoc.exists()) {
      return quizDoc.data();
    }
    return undefined;
  };

  static fetchAboutUsData = async () => {
    const docRef = doc(fStore, "AboutUsData", "Data");
    const aboutUsDoc = await getDoc(docRef);

    if (aboutUsDoc.exists()) {
      return aboutUsDoc.data();
    }
    return undefined;
  };

  static fetchChallengesData = async (currentName: string) => {
    const docRef = doc(fStore, "Challenges", currentName);
    const challengesDoc = await getDoc(docRef);

    if (challengesDoc.exists()) {
      return challengesDoc.data();
    }
    return undefined;
  };

  static fetchChallengesList = async () => {
    const querySnapshot = await getDocs(collection(fStore, "ChallengesData"));

    if (!querySnapshot) return undefined;
    return querySnapshot.docs.map((each) => each.data());
  };
}

export default FirebaseAPI;
