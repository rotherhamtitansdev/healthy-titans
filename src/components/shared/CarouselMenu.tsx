import React, { useState } from "react";
import Carousel from "react-multi-carousel";
import "react-multi-carousel/lib/styles.css";
import { MenuCardProps } from "../../models/MenuCardProps";
import MenuCard from "./MenuCard";

const CarouselMenu = (props: { cards: MenuCardProps[] }) => {
  const [isMoving, setIsMoving] = useState(false);

  const responsive = {
    desktop: {
      breakpoint: { max: 3000, min: 1400 },
      items: 4,
      partialVisibilityGutter: 40,
    },
    smallDesktop: {
      breakpoint: { max: 1400, min: 1024 },
      items: 3,
      partialVisibilityGutter: 50,
    },
    tablet: {
      breakpoint: { max: 1024, min: 800 },
      items: 2,
      partialVisibilityGutter: 100,
    },
    largeMobile: {
      breakpoint: { max: 800, min: 600 },
      items: 2,
      partialVisibilityGutter: 30,
    },
    mobile: {
      breakpoint: { max: 600, min: 464 },
      items: 1,
      partialVisibilityGutter: 200,
    },
    smallMobile: {
      breakpoint: { max: 464, min: 0 },
      items: 1,
      partialVisibilityGutter: 100,
    },
  };

  const cardList = props.cards.map((item) => (
    <MenuCard
      key={item.key}
      name={item.name}
      description={item.description}
      img={item.img}
      path={item.path}
      externalPath={item.externalPath}
      disableOnClick={isMoving}
      firebaseName={item.firebaseName}
    />
  ));

  return (
    <Carousel
      responsive={responsive}
      partialVisible
      beforeChange={() => setIsMoving(true)}
      afterChange={() => setIsMoving(false)}
      itemClass="p-[1rem]"
      renderArrowsWhenDisabled
    >
      {cardList}
    </Carousel>
  );
};
export default CarouselMenu;
