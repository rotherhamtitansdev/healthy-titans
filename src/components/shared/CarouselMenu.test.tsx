import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import React from "react";
import CarouselMenu from "./CarouselMenu";
import { HomePageCarouselResponsiveConfig } from "../../config/CarouselConfig";
import FirebaseAPI from "../../api/FirebaseAPI";
import { MenuCardProps } from "../../models/MenuCardProps";

const MockMenuCards = [
  {
    key: 0,
    name: "Mock Card 1",
    path: "/mock-path-1",
    firebaseName: "MockImages/mock-card-1.jpg",
  },
  {
    key: 1,
    name: "Mock Card 2",
    path: "/mock-path-2",
    firebaseName: "MockImages/mock-card-2.jpg",
  },
  {
    key: 2,
    name: "Mock Card 3",
    path: "/mock-path-3",
    firebaseName: "MockImages/mock-card-3.jpg",
  },
  {
    key: 3,
    name: "Mock Card 4",
    path: "/mock-path-4",
    firebaseName: "MockImages/mock-card-4.jpg",
  },
  {
    key: 4,
    name: "Mock Card 5",
    path: "/mock-path-5",
    firebaseName: "MockImages/mock-card-5.jpg",
  },
] as MenuCardProps[];

beforeEach(() => {
  jest.spyOn(FirebaseAPI, "fetchImages").mockResolvedValue("fruit.jpg");

  Object.defineProperties(window.HTMLElement.prototype, {
    offsetWidth: {
      get: () => 100,
    },
  });
});

describe("Carousel Menu", () => {
  test("component renders", async () => {
    const { asFragment } = render(
      <CarouselMenu
        cards={MockMenuCards}
        config={HomePageCarouselResponsiveConfig}
        renderArrowsWhenDisabled
      />,
      { wrapper: MemoryRouter }
    );

    expect(
      await screen.findByRole("img", {
        name: "Mock Card 1",
      })
    ).toBeVisible();

    // check if the div with the test id is in the document
    expect(screen.getByTestId("Mock Card 1")).toBeVisible();

    // check image source is the same as our mock
    expect(
      screen.getByRole("img", {
        name: "Mock Card 1",
      })
    ).toHaveAttribute("src", "fruit.jpg");

    expect(
      await screen.findByRole("img", {
        name: "Mock Card 1",
      })
    ).toHaveAttribute("alt", "Mock Card 1");

    expect(asFragment()).toMatchSnapshot();
  });

  test("component renders with no cards", async () => {
    render(
      <CarouselMenu
        cards={[]}
        config={HomePageCarouselResponsiveConfig}
        renderArrowsWhenDisabled
      />,
      { wrapper: MemoryRouter }
    );

    expect(
      screen.queryByRole("img", {
        name: "Mock Card 1",
      })
    ).not.toBeInTheDocument();
  });

  test("component renders with both arrows", async () => {
    render(
      <CarouselMenu
        cards={MockMenuCards}
        config={HomePageCarouselResponsiveConfig}
        renderArrowsWhenDisabled
      />,
      { wrapper: MemoryRouter }
    );

    expect(
      screen.getByRole("button", {
        name: "Go to previous slide",
      })
    ).toBeVisible();

    expect(
      screen.getByRole("button", {
        name: "Go to next slide",
      })
    ).toBeVisible();
  });

  test("component renders with only 1 arrow", async () => {
    render(
      <CarouselMenu
        cards={MockMenuCards}
        config={HomePageCarouselResponsiveConfig}
        renderArrowsWhenDisabled={false}
      />,
      { wrapper: MemoryRouter }
    );

    expect(
      screen.queryByRole("button", {
        name: "Go to previous slide",
      })
    ).not.toBeInTheDocument();

    expect(
      screen.getByRole("button", {
        name: "Go to next slide",
      })
    ).toBeVisible();
  });
});
