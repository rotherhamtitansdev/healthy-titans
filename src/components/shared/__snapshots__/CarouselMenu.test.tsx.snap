// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Menu component renders 1`] = `
<DocumentFragment>
  <div
    class="react-multi-carousel-list  "
    dir="ltr"
  >
    <ul
      class="react-multi-carousel-track "
      style="transition: none; overflow: unset; transform: translate3d(0px,0,0);"
    >
      <li
        aria-hidden="false"
        class="react-multi-carousel-item react-multi-carousel-item--active px-[1rem] py-[1.5rem]"
        data-index="0"
        style="flex-basis: auto; position: relative; width: -50px;"
      >
        <div
          class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg w-60 h-60 sm:w-64 sm:h-64 xl:w-72 xl:h-72 2xl:w-80 2xl:h-80 3xl:w-96 3xl:h-96 hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
          data-testid="Mock Card 1"
          role="button"
          tabindex="0"
        >
          <div
            class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
          >
            <div
              class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
            >
              <div
                class="px-[2.75rem] md:px-[5rem] pb-[0.5rem] md:pb-[1rem] min-w-[89px] min-h-[81px] 2xl:mt-5"
              >
                <img
                  alt=""
                  class="mx-auto undefined"
                  draggable="false"
                  src="fruit.jpg"
                />
              </div>
            </div>
            <div
              class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
            >
              Mock Card 1
            </div>
          </div>
        </div>
      </li>
      <li
        aria-hidden="false"
        class="react-multi-carousel-item react-multi-carousel-item--active px-[1rem] py-[1.5rem]"
        data-index="1"
        style="flex-basis: auto; position: relative; width: -50px;"
      >
        <div
          class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg w-60 h-60 sm:w-64 sm:h-64 xl:w-72 xl:h-72 2xl:w-80 2xl:h-80 3xl:w-96 3xl:h-96 hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
          data-testid="Mock Card 2"
          role="button"
          tabindex="0"
        >
          <div
            class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
          >
            <div
              class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
            >
              <div
                class="px-[2.75rem] md:px-[5rem] pb-[0.5rem] md:pb-[1rem] min-w-[89px] min-h-[81px] 2xl:mt-5"
              >
                <img
                  alt=""
                  class="mx-auto undefined"
                  draggable="false"
                  src="fruit.jpg"
                />
              </div>
            </div>
            <div
              class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
            >
              Mock Card 2
            </div>
          </div>
        </div>
      </li>
      <li
        aria-hidden="true"
        class="react-multi-carousel-item  px-[1rem] py-[1.5rem]"
        data-index="2"
        style="flex-basis: auto; position: relative; width: -50px;"
      >
        <div
          class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg w-60 h-60 sm:w-64 sm:h-64 xl:w-72 xl:h-72 2xl:w-80 2xl:h-80 3xl:w-96 3xl:h-96 hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
          data-testid="Mock Card 3"
          role="button"
          tabindex="0"
        >
          <div
            class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
          >
            <div
              class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
            >
              <div
                class="px-[2.75rem] md:px-[5rem] pb-[0.5rem] md:pb-[1rem] min-w-[89px] min-h-[81px] 2xl:mt-5"
              >
                <img
                  alt=""
                  class="mx-auto undefined"
                  draggable="false"
                  src="fruit.jpg"
                />
              </div>
            </div>
            <div
              class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
            >
              Mock Card 3
            </div>
          </div>
        </div>
      </li>
      <li
        aria-hidden="true"
        class="react-multi-carousel-item  px-[1rem] py-[1.5rem]"
        data-index="3"
        style="flex-basis: auto; position: relative; width: -50px;"
      >
        <div
          class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg w-60 h-60 sm:w-64 sm:h-64 xl:w-72 xl:h-72 2xl:w-80 2xl:h-80 3xl:w-96 3xl:h-96 hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
          data-testid="Mock Card 4"
          role="button"
          tabindex="0"
        >
          <div
            class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
          >
            <div
              class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
            >
              <div
                class="px-[2.75rem] md:px-[5rem] pb-[0.5rem] md:pb-[1rem] min-w-[89px] min-h-[81px] 2xl:mt-5"
              >
                <img
                  alt=""
                  class="mx-auto undefined"
                  draggable="false"
                  src="fruit.jpg"
                />
              </div>
            </div>
            <div
              class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
            >
              Mock Card 4
            </div>
          </div>
        </div>
      </li>
      <li
        aria-hidden="true"
        class="react-multi-carousel-item  px-[1rem] py-[1.5rem]"
        data-index="4"
        style="flex-basis: auto; position: relative; width: -50px;"
      >
        <div
          class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg w-60 h-60 sm:w-64 sm:h-64 xl:w-72 xl:h-72 2xl:w-80 2xl:h-80 3xl:w-96 3xl:h-96 hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
          data-testid="Mock Card 5"
          role="button"
          tabindex="0"
        >
          <div
            class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
          >
            <div
              class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
            >
              <div
                class="px-[2.75rem] md:px-[5rem] pb-[0.5rem] md:pb-[1rem] min-w-[89px] min-h-[81px] 2xl:mt-5"
              >
                <img
                  alt=""
                  class="mx-auto undefined"
                  draggable="false"
                  src="fruit.jpg"
                />
              </div>
            </div>
            <div
              class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
            >
              Mock Card 5
            </div>
          </div>
        </div>
      </li>
    </ul>
    <button
      aria-label="Go to previous slide"
      class="react-multiple-carousel__arrow react-multiple-carousel__arrow--left "
      disabled=""
      type="button"
    />
    <button
      aria-label="Go to next slide"
      class="react-multiple-carousel__arrow react-multiple-carousel__arrow--right "
      type="button"
    />
  </div>
</DocumentFragment>
`;
